network/dag/tree/tree.go

```diff
@@ -0,0 +1,313 @@
+	Add(other Data) error
+	// Subtract other Data from this one. Returns an error if the underlying datastructures are incompatible.
+	Subtract(other Data) error
+	encoding.BinaryMarshaler
+	encoding.BinaryUnmarshaler
+}
+
+type Tree interface {
+	// Insert a transaction reference at the specified clock value. The result of inserting the same ref multiple times is undefined.
+	Insert(ref hash.SHA256Hash, clock uint32) error
+	// GetRoot returns the accumulated Data for the entire tree
+	GetRoot() Data
+	// GetZeroTo returns the LC value closest to the requested clock value together with Data of the same range.
+	// The LC value closest to requested clock is defined as the lowest of:
+	// 	- highest LC known to the Tree
+	// 	- highest LC of the leaf that clock is on: ceil(clock/leafSize)*leafSize - 1
+	GetZeroTo(clock uint32) (Data, uint32)
+	// DropLeaves shrinks the tree by dropping all leaves. The parent of a leaf will become the new leaf.
+	DropLeaves()
+	// GetUpdate return the leaves that have been orphaned or updated since the last call to ResetUpdate.
+	// dirty and orphaned are mutually exclusive.
+	GetUpdate() (dirty map[uint32][]byte, orphaned []uint32, err error)
+	// ResetUpdate forgets all currently tracked changes.
+	ResetUpdate()
+	// Load builds a tree from binary leaf data. The keys in leaves correspond to a node's split value.
+	Load(leaves map[uint32][]byte) error
+}
+
+/*
+tree creates a binary tree, where the leaves contain Data over a fixed range of Lamport Clock (uint32) values.
+	- The Data of the parent node is the sum of that of its children. Thus root contains the sum of all Data in the tree.
+	- The value that splits a node into its children is used as a nodeID since it is unique, even after tree resizing.
+	- Since the leaves are of fixed size, a new root is created when added something to a clock outside the current root range.
+	- Whenever a new branch is created, a string of left Nodes is created all the way down to the leaf.
+	- tree is not thread-safe. Since the tree is agnostic to its content, great care must be taken to prevent adding the same data more than once.
+*/
+type tree struct {
+	Depth    int    `json:"depth"`
+	MaxSize  uint32 `json:"max_size"`
+	LeafSize uint32 `json:"leaf_size"`
+	Root     *node  `json:"root"`
+	// storage
+	prototype      Data
+	dirtyLeaves    map[uint32]*node
+	orphanedLeaves map[uint32]struct{}
+}
+
+// New creates a new tree with the given leafSize and of the same type of Data as the prototype.
+func New(prototype Data, leafSize uint32) *tree {
+	tr := &tree{prototype: prototype.New()}
+	tr.resetDefaults(leafSize)
+	return tr
+}
+
+func (t *tree) resetDefaults(leafSize uint32) {
+	t.Root = newNode(leafSize/2, leafSize, t.prototype.New())
+	t.LeafSize = leafSize
+	t.MaxSize = leafSize
+	t.Depth = 0
+	t.dirtyLeaves = map[uint32]*node{t.Root.SplitLC: t.Root}
+	t.orphanedLeaves = nil
+}
+
+func (t *tree) Load(leaves map[uint32][]byte) error {
```

Comment:
Maybe I misunderstood its value, what if the key (`uint32`) does not increase regularly? Due to a bug/data corruption in storage?